client/verta/verta/_tracking/experimentrun.py

```diff
@@ -1749,21 +1860,84 @@ def log_requirements(self, requirements, overwrite=False):
         """
         if isinstance(requirements, six.string_types):
             with open(requirements, 'r') as f:
-                requirements = _pip_requirements_utils.clean_reqs_file_lines(f.readlines())
+                requirements = _pip_requirements_utils.clean_reqs_file_lines(
+                    f.readlines())
         elif (isinstance(requirements, list)
               and all(isinstance(req, six.string_types) for req in requirements)):
             requirements = copy.copy(requirements)
         else:
-            raise TypeError("`requirements` must be either str or list of str, not {}".format(type(requirements)))
+            raise TypeError("`requirements` must be either str or list of str, not {}".format(
+                type(requirements)))
 
         _pip_requirements_utils.process_requirements(requirements)
 
         if self._conf.debug:
             print("[DEBUG] requirements are:")
             print(requirements)
+        python_env = Python(copy.copy(requirements))
+        requirements = six.BytesIO(six.ensure_binary(
+            '\n'.join(requirements)))  # as file-like
+        self._log_artifact("requirements.txt", requirements,
+                           _CommonCommonService.ArtifactTypeEnum.BLOB, 'txt', overwrite=overwrite)
+        try:
+            self.log_environment(python_env, overwrite=overwrite)
+        except ValueError as ve:
+            six.raise_from(ve, None)
+        except TypeError as te:
+            six.raise_from(te, None)
+        except requests.HTTPError as he:
+            pass  # Note: HTTP errors are silently ignored for compatibility with older services
+
+    def log_environment(self, env, overwrite=False):
+        """
+        Logs a Python environment to this Experiment Run.
+
+        Parameters
+        ----------
+        env : :class:`~verta.environment.Python`
+            Environment to log.
+        overwrite : bool, default False
+            Whether to allow overwriting an existing artifact with key `key`.
+
+        """
+        if not isinstance(env, Python):
+            raise TypeError(
+                "`env` must be of type Python, not {}".format(type(env)))
+
+        if self.has_environment and not overwrite:
+            raise ValueError(
+                "environment already exists; consider setting overwrite=True")
+
+        # need to call .environment on the proto object because ._as_proto produces
+        # ai.verta.modeldb.versioning.Blob, not ai.verta.modeldb.versioning.EnvironmentBlob
+        msg = _ExperimentRunService.LogEnvironment(
+            id=self.id, environment=env._as_proto().environment)
+        data = _utils.proto_to_json(msg)
+        response = _utils.make_request("POST",
+                                       "{}://{}/api/v1/modeldb/experiment-run/logEnvironment".format(
+                                           self._conn.scheme, self._conn.socket),
+                                       self._conn, json=data)
```

Comment:
Not required, but the new-and-improved idiomatic way to write this would be
```suggestion
        response = self._conn.make_proto_request(
            "POST", "/api/v1/modeldb/experiment-run/logEnvironment", body=msg,
        )
```
The older `_utils.make_request()` calls need to be cleaned up 😓